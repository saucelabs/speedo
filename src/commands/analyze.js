import ora from 'ora'
import yargs from 'yargs'
import SauceLabs from 'saucelabs'

import { getMetricParams, getJobUrl, analyzeReport, waitFor } from '../utils'
import {
    ERROR_MISSING_CREDENTIALS, ANALYZE_CLI_PARAMS, PERFORMANCE_METRICS
} from '../constants'

export const command = 'analyze [params...] <jobName>'
export const desc = 'Analyze results of prerun performance tests.'
export const builder = ANALYZE_CLI_PARAMS

export const handler = async (argv) => {
    const username = process.env.SAUCE_USERNAME || argv.user
    const accessKey = process.env.SAUCE_ACCESS_KEY || argv.key
    const metrics = getMetricParams(argv)

    /**
     * check if username and access key are available
     */
    if (!username || !accessKey) {
        yargs.showHelp()
        // eslint-disable-next-line no-console
        console.error(ERROR_MISSING_CREDENTIALS)
        return process.exit(1)
    }

    const user = new SauceLabs(username, accessKey)
    const status = ora(`Fetch job "${argv.jobName}" from ${username}`).start()

    /**
     * fetch job
     */
    let job
    try {
        job = (await user.listJobs(
            username,
            { name: argv.jobName, limit: 1 }
        )).jobs.pop()

        /**
         * fail if no job can be found
         */
        if (!job) {
            throw new Error('job not found')
        }

        /**
         * error out if job didn't complete
         */
        if (job.error) {
            throw new Error('job failed or did\'t complete')
        }

        status.succeed()
    } catch (e) {
        status.fail(`Couldn't fetch job with name "${argv.jobName}": ${e.stack}`)
        return process.exit(1)
    }

    job.name = argv.jobName

    status.start('Analyze performance of job')
    const jobResult = {
        id: job.id,
        name: job.name,
        url: getJobUrl(argv, job.id),
        results: []
    }

    try {
        const performanceMetrics = await waitFor(
            () => user.getPerformanceMetrics(job.id),
            /* istanbul ignore next */
            (performanceMetrics) => performanceMetrics.items.length !== 0
        )

        /**
         * filter by page url if given
         */
        if (argv.pageUrl) {
            performanceMetrics.items = performanceMetrics.items.filter(
                (perfMetric) => perfMetric.page_url === argv.pageUrl)
        }

        /**
         * filter by order index if given
         */
        if (typeof argv.orderIndex === 'number') {
            performanceMetrics.items = performanceMetrics.items.filter(
                (perfMetric) => perfMetric.order_index === argv.orderIndex)
        }

        for (const pageLoadMetric of performanceMetrics.items) {
            const results = {}
            const baselineHistory = await user.getBaselineHistory(job.id, {
                metricNames: PERFORMANCE_METRICS,
                orderIndex: pageLoadMetric.order_index
            })

            for (const [metricName, baseline] of Object.entries(baselineHistory)) {
                const capturedValue = pageLoadMetric.metric_data[metricName]
                const result = {
                    metric: metricName,
                    passed: true,
                    value: capturedValue || 0,
                    baseline: baseline[0]
                }

                if (metrics.includes(metricName) && (baseline[0].u < capturedValue || baseline[0].l > capturedValue)) {
                    result.passed = false
                }

                results[metricName] = result
            }

            jobResult.results.push({
                orderIndex: pageLoadMetric.order_index,
                url:  pageLoadMetric.page_url,
                passed: !Object.values(results).find((r) => !r.passed),
                metrics: results
            })
        }

        jobResult.passed = Object.values(jobResult.results).find((r) => !r.passed)
        status.succeed()
    } catch (e) {
        status.fail(`Couldn't fetch performance results: ${e.stack}`)
        return process.exit(1)
    }

    analyzeReport(jobResult, metrics)
    process.exit(jobResult.passed ? 1 : 0)
}
