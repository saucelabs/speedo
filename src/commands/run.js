import fs from 'fs'
import tmp from 'tmp'
import ora from 'ora'
import path from 'path'
import yargs from 'yargs'
import ordinal from 'ordinal'
import SauceLabs from 'saucelabs'
import { camelCase } from 'change-case'

import runPerformanceTest from '../runner'
import {
    printResult, waitFor, getMetricParams, getJobUrl,
    getJobName, getThrottleNetworkParam, getDeviceClassFromBenchmark,
    startTunnel, getConfig, getLigthouseReportUrl, prepareBudgetData,
    getBudgetMetrics, getJankinessParam, printJankinessResult
} from '../utils'
import {
    ERROR_MISSING_CREDENTIALS, REQUIRED_TESTS_FOR_BASELINE_COUNT,
    RUN_CLI_PARAMS, TUNNEL_SHUTDOWN_TIMEOUT, JANKINESS_METRIC
} from '../constants'

export const command = 'run [params...] <site>'
export const desc = 'Run performance tests on any website.'
export const builder = RUN_CLI_PARAMS

export const handler = async (argv) => {
    const config = getConfig(argv)
    const username = config.user || process.env.SAUCE_USERNAME
    const accessKey = config.key || process.env.SAUCE_ACCESS_KEY
    const jobName = getJobName(config)
    const buildName = config.build || `${jobName} - ${(new Date()).toString()}`
    const budget = config ? config.budget : null
    const metrics = budget ? getBudgetMetrics(budget) : getMetricParams(config)
    const jankinessScore = getJankinessParam(argv, budget)
    /**
     * check if username and access key are available
     */
    if (!username || !accessKey) {
        yargs.showHelp()
        // eslint-disable-next-line no-console
        console.error(ERROR_MISSING_CREDENTIALS)
        return process.exit(1)
    }

    const status = ora(`Start performance test run with user ${username} on page ${config.site}...`).start()

    const logDir = config.logDir
        ? path.resolve(process.cwd(), config.logDir)
        : tmp.dirSync().name

    /**
     * check if job already exists
     */
    const user = new SauceLabs({
        user: username,
        key: accessKey,
        region: config.region
    })

    /**
     * find if job already exists
     */
    let job = null
    try {
        job = await user.listJobs(
            username,
            { name: jobName, limit: 10 }
        )
        status.succeed()
    } catch (e) {
        status.fail(`Couldn't fetch job: ${e.stack}`)
        return process.exit(1)
    }

    /**
     * start Sauce Connect if not done by user
     */
    let tunnelProcess
    if (config.tunnelIdentifier) {
        status.start(`Checking for Sauce Connect tunnel with identifier "${config.tunnelIdentifier}"`)
        try {
            tunnelProcess = await startTunnel(user, accessKey, logDir, config.tunnelIdentifier)

            if (tunnelProcess) {
                status.text = `Started Sauce Connect tunnel with identifier "${config.tunnelIdentifier}"`
            }

            status.succeed()
        } catch (err) {
            status.fail(`Problem setting up Sauce Connect: ${err.stack}`)
            return process.exit(1)
        }
    }

    /**
     * create baseline if not enough tests have been executed
     */
    if (job.jobs.length < REQUIRED_TESTS_FOR_BASELINE_COUNT) {
        status.start(`Couldn't find baseline for job with name "${jobName}", creating baseline...`)

        const testCnt = REQUIRED_TESTS_FOR_BASELINE_COUNT - job.jobs.length
        await Promise.all([...Array(testCnt)].map(
            () => runPerformanceTest(username, accessKey, config, jobName, undefined, logDir)))
        status.succeed()
    }

    /**
     * run single test
     */
    status.start('Run performance test...')
    let { result, sessionId, benchmark, userAgent, jankinessResult } = await runPerformanceTest(
        username, accessKey, config, jobName, buildName, logDir, jankinessScore)

    /**
     * retry performance test
     */
    const retriedJobs = []
    if (result.result !== 'pass' && config.retry) {
        for (let retry = 1; retry <= config.retry; ++retry) {
            retriedJobs.push(sessionId)
            status.text = `Run performance test (${ordinal(retry)} retry)...`

            const retriedResult = await runPerformanceTest(
                username, accessKey, config, jobName, buildName, logDir, jankinessScore)

            result = retriedResult.result
            sessionId = retriedResult.sessionId
            jankinessResult = retriedResult.jankinessResult

            /**
             * continue command if job has finally passed
             */
            if (result.result === 'pass') {
                break
            }
        }
    }

    status.succeed()

    /**
     * wait until job completes
     */
    try {
        status.start('Wait for job to finish...')
        await waitFor(
            () => user.getJob(username, sessionId),
            /* istanbul ignore next */
            (jobDetails) => jobDetails.status === 'complete'
        )
        status.succeed()
    } catch (e) {
        status.fail(e.message)
        return process.exit(1)
    }

    /**
     * download performance logs
     */
    let performanceLog
    try {
        status.start('Download performance logs...')
        const perfMetrics = await waitFor(
            () => user.getPerformanceMetrics(sessionId),
            /* istanbul ignore next */
            (performanceMetrics) => performanceMetrics.items.length !== 0,
            'Couldn\'t receive any performance metrics'
        )
        performanceLog = perfMetrics.items.map((item) => ({
            sessionId: item.job_id,
            url: item.page_url,
            orderIndex: item.order_index,
            loaderId: item.loader_id,
            metrics: Object.entries(item.metric_data).reduce((obj, [metricName, metricValue]) => {
                obj[camelCase(metricName)] = metricValue
                return obj
            }, {})
        }))

        /**
         * store data log dir
         */
        fs.writeFileSync(
            path.join(logDir, 'performance.json'),
            JSON.stringify(performanceLog, null, 4)
        )

        status.succeed()
    } catch (e) {
        status.fail(`Couldn't download performance results due to: ${e.stack}`)
        return process.exit(1)
    }

    /**
     * download trace file if requested
     */
    const loaderId = performanceLog[0].loaderId
    if (config.traceLogs) {
        status.start('Download trace logs...')
        try {
            await user.downloadJobAsset(
                sessionId,
                `_tracelog_${loaderId}.json.gz`,
                path.join(logDir, 'trace.json'))
            status.succeed()
        } catch (e) {
            status.fail(`Couldn't fetch trace logs: ${e.stack}`)
            status.stopAndPersist({ text: 'continuing ...' })
        }
    }

    /**
     * update job if performance check fails
     */
    try {
        status.start('Updating job status...')
        await user.updateJob(username, sessionId, { passed: result.result === 'pass' })
        status.succeed()
    } catch (e) {
        status.fail(`Couldn't update job due to: ${e.stack}`)
        status.stopAndPersist({ text: 'continuing ...' })
    }

    /**
     * stop tunnel if one was created in the beginning
     */
    if (tunnelProcess) {
        status.start('Stopping Sauce Connect tunnel...')
        await new Promise((resolve) => {
            const shutdownTimeout = setTimeout(() => {
                status.warn('Sauce Connect shutdown timedout, you may still have tunnels lingering around')
                return resolve()
            }, TUNNEL_SHUTDOWN_TIMEOUT)

            return tunnelProcess.close(() => {
                clearTimeout(shutdownTimeout)
                status.succeed()
                resolve()
            })
        })
    }

    status.stopAndPersist({
        text: `Stored performance logs in ${logDir}`,
        symbol: 'üìÉ'
    })

    if (budget) {
        status.stopAndPersist({
            text: 'Asserting against performance budget of config file',
            symbol: 'üßê '
        })
        const preparedBudget = prepareBudgetData(budget)
        for (const [metric, [{ l: lowerLimit, u: upperLimit }]] of Object.entries(preparedBudget)) {
            const actual = performanceLog[0].metrics[metric]
            /**
             * don't fail if - metric is within budget
             */
            if (actual <= upperLimit && actual >= lowerLimit || metric === JANKINESS_METRIC) {
                continue
            }

            result.result = 'failed'
            result.details = result.details || {}
            result.details[metric] = {
                actual,
                upperLimit,
                lowerLimit,
            }
        }
    }

    if (jankinessScore) {
        status.stopAndPersist({
            text: 'Asserting against jankinessScore',
            symbol: 'üßê '
        })

        if (!jankinessResult) {
            status.fail('Couldn\'t capture jankiness result')
        } else {
            const capturedJankinessScore = Math.round(jankinessResult.score * 100)
            const [{ l: lowerLimit, u: upperLimit }] = jankinessScore[JANKINESS_METRIC]
            if (capturedJankinessScore > upperLimit || capturedJankinessScore < lowerLimit) {
                result.result = 'failed'
                result.details = result.details || {}
                result.details[JANKINESS_METRIC] = {
                    actual: capturedJankinessScore,
                    upperLimit,
                    lowerLimit,
                }
            }
            printJankinessResult(jankinessResult)
        }
    }

    printResult(result, performanceLog[0], metrics, config)

    const networkCondition = getThrottleNetworkParam(config)
    const runtimeSettings = [
        `- Network Throttling: ${networkCondition}`,
        `- CPU Throttling: ${config.throttleCpu}x`,
        `- CPU/Memory Power: ${benchmark} (${getDeviceClassFromBenchmark(benchmark)})`,
        `- User Agent: ${userAgent}`
    ]
    status.stopAndPersist({
        text: `Runtime settings:\n${runtimeSettings.join('\n')}\n`,
        symbol: '‚öôÔ∏è '
    })

    status.stopAndPersist({
        text: `Check out job at ${getJobUrl(config, sessionId)}`,
        symbol: 'üëÄ'
    })

    status.stopAndPersist({
        text: `Check out Lighthouse Report at ${getLigthouseReportUrl(config, sessionId, loaderId)}`,
        symbol: 'üìî'
    })

    /**
     * displayed retried jobs that failed
     */
    if (retriedJobs.length) {
        console.log( // eslint-disable-line no-console
            '\nFailed Performance Tests that were rerun:\n' +
            retriedJobs.map(id => getJobUrl(config, id)).join('\n')
        )
    }

    process.exit(result.result === 'pass' ? 0 : 1)
}
