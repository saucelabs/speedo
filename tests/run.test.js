import ora from 'ora'
import yargs from 'yargs'
import { fixtures, lastInstance, resetSauceLabsFixtures } from 'saucelabs'

import { handler } from '../src/commands/run'
import runPerformanceTest from '../src/runner'
import {
    waitFor, getMetricParams, getJobUrl, startTunnel, getBudgetMetrics,
    getConfig, getLigthouseReportUrl, prepareBudgetData, printResult,
    printJankinessResult, getJankinessParam,
} from '../src/utils'
import { JANKINESS_RESULT } from './__fixtures__/jankiness'

jest.mock('../src/runner')
jest.mock('../src/utils')
jest.mock('../src/constants', () => ({
    TUNNEL_SHUTDOWN_TIMEOUT: 100,
    REQUIRED_TESTS_FOR_BASELINE_COUNT: 10,
    JANKINESS_METRIC: 'jankiness'
}))
jest.mock('fs')

beforeEach(() => {
    jest.spyOn(process, 'exit').mockImplementation(() => {})
    delete process.env.SAUCE_USERNAME
    delete process.env.SAUCE_ACCESS_KEY
    getConfig.mockImplementation((argv) => argv)
    getMetricParams.mockImplementation(() => ['speedIndex', 'pageWeight'])
    getJobUrl.mockImplementation(() => 'https://saucelabs.com/performance/foobar/0')
    getLigthouseReportUrl.mockImplementation(() => 'https://eds.us-west-1.saucelabs.com/foobar/performance/barfoo/lhr.html')
    waitFor.mockImplementation((condition) => condition())
    runPerformanceTest.mockImplementation(() => ({
        sessionId: 'foobar123',
        result: { result: 'pass' },
        benchmark: 1234,
        userAgent: 'chrome'
    }))
    printResult.mockImplementation((argv) => argv)
    printJankinessResult.mockImplementation((argv) => argv)
    prepareBudgetData.mockImplementation(({ load }) => ( load === 1000 ? {
        load: [{ l: 0, u: 1000 }],
        speedIndex: [{ l: 500, u: 1000 }]
    } : {
        load: [{ l: 0, u: 2000 }],
        speedIndex: [{ l: 1000, u: 1500 }]
    }))
    getJankinessParam.mockImplementation(({ jankiness }) => {
        if (!jankiness) {
            return null
        }
        return jankiness === 50 ?
            { jankiness: [{ l: 50, u: 100 }] } :
            { jankiness: [{ l: 90, u: 100 }] }
    })
    getBudgetMetrics.mockImplementation((budget) => Object.keys(budget))
})

test('run should fail if no auth is provided', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {})
    await handler({})
    // eslint-disable-next-line no-console
    expect(console.error).toBeCalledTimes(1)
    expect(process.exit).toBeCalledWith(1)
    expect(yargs.showHelp).toBeCalledTimes(1)
    // eslint-disable-next-line no-console
    console.error.mockRestore()
})

test('fails when jobs can not be fetched', async () => {
    fixtures.listJobs = Promise.reject(new Error('buh'))
    await handler({ user: 'foo', key: 'bar', metric: ['load', 'speedIndex'] })
    expect(ora().fail).toBeCalledTimes(1)
    expect(ora().fail.mock.calls[0][0]).toContain('Couldn\'t fetch job')
    expect(process.exit).toBeCalledTimes(1)
})

test('should create a new baseline if run with no jobs', async () => {
    const opts = {
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        metric: ['load', 'speedIndex'],
        name: 'foobar'
    }
    fixtures.listJobs = { jobs: [] }
    await handler(opts)
    expect(process.exit).toBeCalledWith(0)
    expect(ora().start)
        .toBeCalledWith('Couldn\'t find baseline for job with name "undefined", creating baseline...')
})

test('should rerun performance tests if they fail', async () => {
    jest.spyOn(console, 'log').mockImplementation(() => {})
    let failures = 0
    runPerformanceTest.mockImplementation(() => {
        if (failures === 3) {
            return {
                sessionId: 'foobarRetried123',
                result: { result: 'pass' },
                benchmark: 1234,
                userAgent: 'chrome'
            }
        }
        ++failures
        return {
            sessionId: 'foobar123',
            result: { result: 'failed' },
            benchmark: 1234,
            userAgent: 'chrome'
        }
    })
    await handler({ user: 'foo', key: 'bar', site: 'mypage', retry: 3 })
    expect(ora().text).toBe('Run performance test (3rd retry)...')
    // eslint-disable-next-line no-console
    expect(console.log).toBeCalledTimes(1)
    expect(process.exit).toBeCalledWith(0)
    // eslint-disable-next-line no-console
    console.log.mockRestore()
})

test('should fail if job never completes', async () => {
    fixtures.getJob = Promise.reject(new Error('buh'))
    await handler({ user: 'foo', key: 'bar', site: 'mypage', metric: ['load', 'speedIndex'] })
    expect(process.exit).toBeCalledTimes(1)
})

test('should fail if performance logs can not be fetched', async () => {
    fixtures.getPerformanceMetrics = Promise.reject(new Error('buhhu'))
    await handler({ user: 'foo', key: 'bar', site: 'mypage', metric: ['load', 'speedIndex'] })
    expect(process.exit).toBeCalledTimes(1)
    expect(ora().fail.mock.calls[0][0]).toContain('Couldn\'t download performance results due to: Error: buhhu')
})

test('should continue if fetching trace logs fails', async () => {
    fixtures.downloadJobAsset = Promise.reject(new Error('buhhu'))
    await handler({ user: 'foo', key: 'bar', site: 'mypage', metric: ['load', 'speedIndex'], traceLogs: true })
    expect(process.exit).toBeCalledWith(0)
    expect(ora().fail.mock.calls[0][0]).toContain('Couldn\'t fetch trace logs: Error: buhhu')
})

test('should store tracelogs if path provided', async () => {
    await handler({
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        metric: ['load', 'speedIndex'],
        traceLogs: true,
        logDir: '/foo/bar'
    })
    expect(lastInstance.downloadJobAsset.mock.calls).toMatchSnapshot()
})

test('should not fail if updating job status fails', async () => {
    fixtures.updateJob = Promise.reject(new Error('ups'))
    await handler({ user: 'foo', key: 'bar', site: 'mypage', metric: ['load', 'speedIndex'] })
    expect(process.exit).toBeCalledWith(0)
    expect(ora().fail.mock.calls[0][0]).toContain('Couldn\'t update job due to: Error: ups')
})

test('should run successfully', async () => {
    await handler({ user: 'foo', key: 'bar', site: 'mypage', metric: ['load', 'speedIndex'] })
    expect(process.exit).toBeCalledWith(0)
    expect(ora().start.mock.calls).toMatchSnapshot()
    expect(ora().stopAndPersist.mock.calls).toMatchSnapshot()
})

test('should fail if captured values are out of budget', async () => {
    await handler({
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        budget: {
            load: 1000,
            speedIndex: [500, 1000]
        }
    })
    expect(printResult.mock.calls[0][0]).toMatchSnapshot()
    expect(process.exit).toBeCalledWith(1)    
})

test('should pass if captured values are within budget', async () => {
    await handler({
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        budget: {
            load: 2000,
            speedIndex: [1000, 1500]
        }
    })
    expect(printResult.mock.calls[0][0]).toMatchSnapshot()
    expect(process.exit).toBeCalledWith(0)    
})

test('should run successfully with tunnels', async () => {
    const opts = {
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        metric: ['load', 'speedIndex'],
        tunnelIdentifier: 'foobar'
    }
    const tunnelMock = {
        close: jest.fn().mockImplementation((cb) => setTimeout(cb, 100))
    }
    startTunnel.mockReturnValue(Promise.resolve(tunnelMock))

    await handler(opts)
    expect(process.exit).toBeCalledWith(0)
    expect(startTunnel).toBeCalledWith(
        expect.any(Object),
        'bar',
        '/some/tmpDir',
        opts.tunnelIdentifier
    )
    expect(tunnelMock.close).toBeCalledTimes(1)
})

test('should fail if tunnel can not be started', async () => {
    const opts = {
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        metric: ['load', 'speedIndex'],
        tunnelIdentifier: 'foobar'
    }
    startTunnel.mockReturnValue(Promise.reject(new Error('ups')))
    await handler(opts)
    expect(ora().fail.mock.calls.pop()[0]).toContain('Problem setting up Sauce Connect')
    expect(process.exit).toBeCalledWith(1)
})

test('should not close tunnel if none was started', async () => {
    const tunnelMock = {
        close: jest.fn().mockImplementation((cb) => setTimeout(cb, 100))
    }
    startTunnel.mockReturnValue(Promise.resolve(null))

    await handler({
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        metric: ['load', 'speedIndex'],
        tunnelIdentifier: 'foobar'
    })
    expect(process.exit).toBeCalledWith(0)
    expect(startTunnel).toBeCalledTimes(1)
    expect(tunnelMock.close).toBeCalledTimes(0)
})

test('should warn if tunnel can not be closed', async () => {
    const tunnelMock = { close: jest.fn() }
    startTunnel.mockReturnValue(Promise.resolve(tunnelMock))

    await handler({
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        metric: ['load', 'speedIndex'],
        tunnelIdentifier: 'foobar'
    })
    expect(process.exit).toBeCalledWith(0)
    expect(startTunnel).toBeCalledTimes(1)
    expect(tunnelMock.close).toBeCalledTimes(1)
    expect(ora().warn.mock.calls[0][0])
        .toContain('Sauce Connect shutdown timedout')
})

test('should fail build if performance test fails', async () => {
    runPerformanceTest.mockImplementation(() => ({
        sessionId: 'foobar123',
        result: { result: 'failed' },
        benchmark: 1234,
        userAgent: 'chrome'
    }))
    await handler({ user: 'foo', key: 'bar', site: 'mypage', metric: ['load', 'speedIndex'] })
    expect(process.exit).toBeCalledWith(1)
})

test('should check jankiness', async () => {
    runPerformanceTest.mockImplementation(() => ({
        sessionId: 'foobar123',
        result: { result: 'pass' },
        benchmark: 1234,
        userAgent: 'chrome',
        jankinessResult: JANKINESS_RESULT
    }))
    await handler({
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        jankiness: 50
    })
    
    expect(printJankinessResult.mock.calls[0][0]).toMatchSnapshot()
    expect(printResult.mock.calls[0][0]).toMatchSnapshot()
    expect(process.exit).toBeCalledWith(0)
})

test('should fail if jankiness score is out of budget', async () => {
    runPerformanceTest.mockImplementation(() => ({
        sessionId: 'foobar123',
        result: { result: 'pass' },
        benchmark: 1234,
        userAgent: 'chrome',
        jankinessResult: JANKINESS_RESULT
    }))
    await handler({
        user: 'foo',
        key: 'bar',
        site: 'mypage',
        jankiness: [90, 100]
    })
    
    expect(printJankinessResult.mock.calls[0][0]).toMatchSnapshot()
    expect(printResult.mock.calls[0][0]).toMatchSnapshot()
    expect(process.exit).toBeCalledWith(1)
})

afterEach(() => {
    process.exit.mockRestore()
    ora.mockClear()
    startTunnel.mockClear()
    ora().fail.mockClear()
    ora().start.mockClear()
    ora().stopAndPersist.mockClear()
    ora().warn.mockClear()
    resetSauceLabsFixtures()
    printResult.mockClear()
    prepareBudgetData.mockClear()
    getBudgetMetrics.mockClear()
    printJankinessResult.mockClear()
    getJankinessParam.mockClear()
})
