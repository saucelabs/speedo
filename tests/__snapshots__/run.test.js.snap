// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should check jankiness 1`] = `
Object {
  "loaderId": "1cfbee70-f4a3-11e9-9108-e7aacbba72d3",
  "score": 0.8813205294283949,
  "timestamp": 1571731701463,
  "type": "scroll",
  "url": "https://www.saucedemo.com/",
  "value": Object {
    "diagnostics": Object {
      "fpsScore": 1,
      "idleDurationScore": 0.9674039580908033,
      "layoutUpdateScore": 1,
      "memoryUsageScore": 0.9936344788683276,
    },
    "metrics": Object {
      "averageFPS": 62.637486770808955,
      "forcedReflowWarningCounts": 0,
      "idleTime": 4986,
      "memoryUsageDiff": 10027708,
      "otherTime": 168,
      "paintingTime": 3,
      "renderingTime": -1,
      "scriptingTime": -2,
      "scrollTime": 5154,
    },
  },
}
`;

exports[`should check jankiness 2`] = `
Object {
  "result": "pass",
}
`;

exports[`should fail if captured values are out of budget 1`] = `
Object {
  "details": Object {
    "load": Object {
      "actual": 1837,
      "lowerLimit": 0,
      "upperLimit": 1000,
    },
    "speedIndex": Object {
      "actual": 1143,
      "lowerLimit": 500,
      "upperLimit": 1000,
    },
  },
  "result": "failed",
}
`;

exports[`should fail if jankiness score is out of budget 1`] = `
Object {
  "loaderId": "1cfbee70-f4a3-11e9-9108-e7aacbba72d3",
  "score": 0.8813205294283949,
  "timestamp": 1571731701463,
  "type": "scroll",
  "url": "https://www.saucedemo.com/",
  "value": Object {
    "diagnostics": Object {
      "fpsScore": 1,
      "idleDurationScore": 0.9674039580908033,
      "layoutUpdateScore": 1,
      "memoryUsageScore": 0.9936344788683276,
    },
    "metrics": Object {
      "averageFPS": 62.637486770808955,
      "forcedReflowWarningCounts": 0,
      "idleTime": 4986,
      "memoryUsageDiff": 10027708,
      "otherTime": 168,
      "paintingTime": 3,
      "renderingTime": -1,
      "scriptingTime": -2,
      "scrollTime": 5154,
    },
  },
}
`;

exports[`should fail if jankiness score is out of budget 2`] = `
Object {
  "details": Object {
    "jankiness": Object {
      "actual": 88,
      "lowerLimit": 90,
      "upperLimit": 100,
    },
  },
  "result": "failed",
}
`;

exports[`should include header when using sauce rest API 1`] = `
Object {
  "User-Agent": "Speedo/1.2.4 (nodejs darwin)",
}
`;

exports[`should pass if captured values are within budget 1`] = `
Object {
  "result": "pass",
}
`;

exports[`should run successfully 1`] = `
Array [
  Array [],
  Array [
    "Run performance test...",
  ],
  Array [
    "Wait for job to finish...",
  ],
  Array [
    "Download performance logs...",
  ],
  Array [
    "Updating job status...",
  ],
]
`;

exports[`should run successfully 2`] = `
Array [
  Array [
    Object {
      "symbol": "üìÉ",
      "text": "Stored performance logs in /some/tmpDir",
    },
  ],
  Array [
    Object {
      "symbol": "‚öôÔ∏è ",
      "text": "Runtime settings:
- Network Throttling: undefined
- CPU Throttling: undefinedx
- CPU/Memory Power: 1234 (undefined)
- User Agent: chrome
",
    },
  ],
  Array [
    Object {
      "symbol": "üëÄ",
      "text": "Check out job at https://saucelabs.com/performance/foobar/0",
    },
  ],
  Array [
    Object {
      "symbol": "üìî",
      "text": "Check out Lighthouse Report at https://api.us-west-1.saucelabs.com/v1/eds/foobar/performance/barfoo/lhr.html",
    },
  ],
]
`;

exports[`should store tracelogs if path provided 1`] = `
Array [
  Array [
    "foobar123",
    "_tracelog_10285BFC9E231DFB7E269641B3CC670.json.gz",
    "/foo/bar/trace.json",
  ],
]
`;
